Added semaphores to synchronize producer and consumer

NeoraptX opened the pull request:
Added semaphores to synchronize consumers with producers

NeoraptX committed:
Added semaphores to synchronize producer and consumer

rockdrigofc commented on MultiThreading/thread_example.cpp:
```diff
@@ -2,16 +2,25 @@
 // dos threads accediendo y/o modificando un recurso compartido int flag, queue, list 
 
 std::queue<int> myQueue;
+sem_t semaphore_full;
```
sem_t is a struct or a class type? @ramsesrago

NeoraptX replied:
yes,. it's a struct

NeoraptX committed:
Added a new mutex for synchronization between consumer and producer
** we have a bug, sometimes the producer produces one more item after the stop flag has been set

NeoraptX committed:
Fixed bug where the queue was not getting drained when the "Stop thread"
was getting called in the middle of execution of producer thread

JuanCodeLab commented on MultiThreading/thread_example.cpp:
```diff
@@ -2,70 +2,108 @@
+    stop = true;
+}
+
+void drainQueue() {
+    // Extraer datos de la cola myQueue     
+    int data = 0;
+    while (!myQueue.empty()) { // para asegurar que se consuman todos los datos despues del stop
+        sem_wait(&semaphore_full);        // Stay here if the buffer is empty
+        sem_wait(&semaphore_mutex);
+        data = myQueue.front();           // Obtiene el elemento "mas viejo" de la cola FIFO
+        myQueue.pop();                    // Remueve el elemento "mas viejo" de la cola
+        std::cout << "CONSUMER: Printing data: " << data 
+                        <<  " now the size of the queue is: " << myQueue.size() << std::endl;
+        sem_post(&semaphore_mutex);
+        sem_post(&semaphore_empty);        // Incrementa el valor de este semaforo
+    }
+}
```
This method only drain Queue but it does not prevent the producer from producing

CONSUMER: Printing data: 32 now the size of the queue is: 1                                                                                                                        
CONSUMER: Printing data: 1c now the size of the queue is: 0                                                                                                                        
**_CONSUMER: Queue empty, goodbye_**                                                                                                                                                     
PRODUCER: Generated random number is: 5d now the queue size is: 1              _**// one more after the stop**_                                                                                                     
CONSUMER: Printing data: 5d now the size of the queue is: 0                            _**// consume one more after the stop**_

The program says goodbye queue empty, but this is not true the sequence continue work after finish the threads 

We have a synchronization problem when we stop, :S 

When start the sequence we have a times for synchronize the work in the threads  whit sleeps but when try stop all threads we don't have synchronization, try stop all threads simultaneously, we need generate this sincronize 

Ihave a propopsal

create two stops one for producers and one for consumers, change value false to true with a difrence of times, first producers and after consumers and you dont need one extra drain, the work of consumer is drain the items in the queue, only we need a synchronize :)

````#  
#include <iostream>
#include <pthread.h>
#include <unistd.h>
#include <queue> // FIFO, pilas son LIFO
#include <semaphore.h>

#define MAX_QUEUE_SIZE    10
// producer-consumer problem
// race conditions
// dos threads accediendo y/o modificando un recurso compartido int flag, queue, list 

std::queue<int> myQueue;
sem_t semaphore_full;
sem_t semaphore_empty;
sem_t semaphore_mutex;
bool stop = false;
bool stop1 = false;

void* stop_producer(void* ptr) {
    sleep(3);
    stop = true;
    sleep(1);
    stop1 = true;
}

void* producer(void* ptr) {
    sleep(2);
    pthread_t thread = pthread_self();
    char threadName[16];
    pthread_getname_np(thread, threadName, sizeof(threadName));
    std::cout << "Hello from " << std::string(threadName) << " with tid: " 
              << "0x" << std::hex << thread <<std::endl;
              
    // aqui comienza el trabajo real del thread
    int randomNumber = 0;
    while (!stop) {
        randomNumber = rand() % 100;    // genera un numero entero aleatorio entre 0 y 100
        sem_wait(&semaphore_empty);     // stay here if the buffer is full, sem = 0, inicialmente vale 10 -> 9
        sem_wait(&semaphore_mutex); 
        myQueue.push(randomNumber);     // Introduce el numero aleatorio generado a la cola
        std::cout << "PRODUCER: Generated random number is: " << randomNumber
                  << " now the queue size is: " << myQueue.size() << std::endl;
        sem_post(&semaphore_mutex); 
        sem_post(&semaphore_full);      // increment the buffer counter for this semaphore
    }
    //sem_post(&semaphore_mutex);
    return NULL;
}

void* consumer(void* ptr) {
    sleep(1);
    pthread_t thread = pthread_self();
    char threadName[16];
    pthread_getname_np(thread, threadName, sizeof(threadName));
    std::cout << "CONSUMER: Hello from " << std::string(threadName) << " with tid: " 
              << "0x" << std::hex << thread << std::endl;
    
    // Extraer datos de la cola myQueue     
    int data = 0;
    while (!stop1) {
        while (!myQueue.empty()) { // para asegurar que se consuman todos los datos despues del stop
            sem_wait(&semaphore_full);        // Stay here if the buffer is empty
            sem_wait(&semaphore_mutex);
            data = myQueue.front();           // Obtiene el elemento "mas viejo" de la cola FIFO
            myQueue.pop();                    // Remueve el elemento "mas viejo" de la cola
            std::cout << "CONSUMER: Printing data: " << data 
                          <<  " now the size of the queue is: " << myQueue.size() << std::endl;
            sem_post(&semaphore_mutex);
            sem_post(&semaphore_empty);        // Incrementa el valor de este semaforo
        }
        // aqui se quedo consumidor
    }
    // la cola esta vacia
    std::cout << "CONSUMER: Queue empty, goodbye " << std::endl;
    return NULL;
}

int main() {
    std::cout << "hello from main" << std::endl;
    //Initialize sempahore
    sem_init(&semaphore_full, 0, 0);
    sem_init(&semaphore_empty, 0, MAX_QUEUE_SIZE);
    sem_init(&semaphore_mutex, 0, 1);
    
    // creamos los trheads
    pthread_t thread1;
    pthread_t thread2;
    pthread_t thread3;
    
    // ejecutamos los threads
    pthread_create(&thread1, NULL, &producer, NULL);
    pthread_create(&thread2, NULL, &consumer, NULL);
    pthread_create(&thread3, NULL, &stop_producer, NULL);
    
    // nombramos los threads
    pthread_setname_np(thread1, "Producer thread");
    pthread_setname_np(thread2, "Consumer thread");
    pthread_setname_np(thread2, "Stop producer");
    
    // esperamos a que los trheads terminen
    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);
    pthread_join(thread3, NULL);
    
    return 0;
}`````

Result in screen 

CONSUMER: Printing data: 31 now the size of the queue is: 1                                                                                                                        
CONSUMER: Printing data: 26 now the size of the queue is: 0                                                                                                                        
CONSUMER: Queue empty, goodbye

JuanCodeLab replied:
Results in screen 

**_CONSUMER: Printing data: 31 now the size of the queue is: 1_**                                                                                                                        
**_CONSUMER: Printing data: 26 now the size of the queue is: 0_**                                                                                                                        
**_CONSUMER: Queue empty, goodbye_**

NeoraptX committed:
Small changes after review with JC:
- Added stop flag for consumer only so that we can stop the consumer only after the producer has finished its execution

rockdrigofc reviewed (approved):
+1
Comments have been clarified. We can proceed with merger to master

NeoraptX reviewed (commented):
- Add Makefile
- Add tests
- Objetify

JuanCodeLab reviewed (approved):
I like the new solution and method reorganization 👍
